indexing-service/src/main/java/io/druid/indexing/firehose/IngestSegmentFirehoseFactory.java

```diff
@@ -286,6 +214,50 @@ public WindowedStorageAdapter apply(final PartitionChunk<DataSegment> input)
     catch (IOException | SegmentLoadingException e) {
       throw Throwables.propagate(e);
     }
+  }
+
+  public static List<String> getUniqueDimensions(
+      List<TimelineObjectHolder<String, DataSegment>> timelineSegments,
+      InputRowParser inputRowParser
+  )
+  {
+    final Set<String> dimSet = timelineSegments
```

Comment:
I know the old code didn't do this, but it would be best to retain the order of dimensions specified in the underlying segments. This is because the user may have chosen a particular dimension sort order in order to optimize queries.

Reply:
Good point. But, I wonder what if two segments have totally different dimension orders. For an extreme example, a segment has 'A', 'B', 'C', 'D' dimensions and another has 'D', 'C', 'B', 'A'.
I think some algorithms like n-way merge can be used for general cases, but I'm not sure about the extreme case.

Reply:
Changed to retain the original order. If segments have different orders, unique dimensions are searched from the recent segment to the oldest one. This is because recent segments are likely to be queried more frequently, and thus the performance should be optimized for recent ones rather than old ones.

Reply:
If two segments have totally different dimension orders, that's fine, we can just pick one or something like that. And yeah going with recent makes sense to me, for another reason as well: if a user changes the order at some point, more recent segments are more likely to have the new desired order, and so it's best to go with them.